### ISSUE TITLE
[Bug] FlashInfer backend drops attn_logit_softcapping: the cap is passed to the deprecated forward() instead of plan(), so gemma-2 and grok-1 run uncapped

### ISSUE THREAD

**Issue by hijohnnylin** [sgl-project/sglang#33915]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

The FlashInfer attention backend never applies `logit_cap`. The cap is read per layer and handed to the
FlashInfer wrapper's deprecated `forward()` API, but `use_logits_soft_cap` is a **compile-time** property
baked into the module by `plan()` / `begin_forward()`, so it cannot be supplied per call — `forward()`
accepts the argument and silently ignores it. Every model that sets a nonzero `logit_cap` therefore runs
with no capping at all on the default backend:

- `gemma2.py:161,180` — `logit_cap = config.attn_logit_softcapping` (50.0 for the gemma-2 family)
- `grok.py:420,429` — `logit_cap = max(getattr(config, "attn_logit_softcapping", 30.0), 0.0)`

(`gemma3_causal.py`, `gemma4_causal.py` and `gemma3n_causal.py` pass `logit_cap=0.0`, so they are
unaffected.) The `triton` backend applies the cap correctly, so the two backends disagree on the same
checkpoint.

This is the mechanism behind [flashinfer#1193](https://github.com/flashinfer-ai/flashinfer/issues/1193)
(*"logit cap is wrong in flashinfer >= 0.2.0"*, filed 2025-06-30 from the SGLang side with GSM8K numbers
for gemma and grok-1, no replies). FlashInfer's kernels apply the cap correctly when `plan()` is told
about it — I verified that directly, below — so the actionable fix is here.

### Where the cap is dropped

`python/sglang/srt/layers/attention/flashinfer_backend.py`, on `main` (`18e6c61`) and on 0.5.16 alike.
Passed at call time:

- `forward_extend` → `prefill_wrapper_paged.forward(..., logits_soft_cap=logits_soft_cap)` (:1318),
  `prefill_wrapper_ragged.forward(...)` (:1353), and both `forward_return_lse(...)` calls (:1372, :1380)
- `forward_decode` → `decode_wrapper.forward(..., logits_soft_cap=layer.logit_cap)` (:1443–1447)

while none of the plan calls mentions it:

- ragged prefill: `wrapper_ragged.begin_forward(qo_indptr, qo_indptr, num_qo_heads, num_kv_heads, head_dim, q_data_type=...)` (:2114)
- paged prefill: `wrapper_paged.begin_forward(...)` (:2189)
- decode: `wrapper.begin_forward(...)` (:1699 and :1718)

The same function already treats `window_left` as plan-time state for the paged wrapper, with the comment
*"selects the module with the per-element window mask compiled in"* (:2185–2187). `logits_soft_cap`
belongs in exactly that category and was missed.

### `window_left` has the same problem on the ragged path

The ragged prefill wrapper is planned without `window_left` (:2114) and given it at `forward()` (:1371,
:1380), so it is dropped by the same mechanism — the second script below shows it. I have not measured a
model-level impact for the window (the prompts I ran are far shorter than any sliding window), but it is
the same bug on the same path.

### 1. No model needed: `forward()` ignores the cap

Three call patterns through the same ragged-prefill wrapper, against an fp32 torch reference computed both
with and without the cap. Shapes and scale mirror `gemma-2-27b` (32 q heads, 16 KV heads, `head_dim=128`,
`sm_scale=1/sqrt(144)`, cap 50), and the gain is set so logits reach the cap:

```python
import math

import flashinfer
import torch

DEV = "cuda"
SEQ, NH, NKV, HD = 64, 32, 16, 128
CAP, SM_SCALE = 50.0, 1.0 / math.sqrt(144.0)

torch.manual_seed(0)
q = (torch.randn(SEQ, NH, HD, device=DEV) * 12).bfloat16()
k = (torch.randn(SEQ, NKV, HD, device=DEV) * 12).bfloat16()
v = torch.randn(SEQ, NKV, HD, device=DEV, dtype=torch.bfloat16)


def reference(*, apply_cap):
    kf = k.float().repeat_interleave(NH // NKV, 1)
    vf = v.float().repeat_interleave(NH // NKV, 1)
    scores = torch.einsum("qhd,khd->hqk", q.float(), kf) * SM_SCALE
    logits = CAP * torch.tanh(scores / CAP) if apply_cap else scores
    mask = torch.triu(torch.ones(SEQ, SEQ, dtype=torch.bool, device=DEV), 1)
    return torch.einsum("hqk,khd->qhd", logits.masked_fill(mask, float("-inf")).softmax(-1), vf)


def cos(a, b):
    a, b = a.float().flatten(), b.float().flatten()
    return float(torch.dot(a, b) / (a.norm() * b.norm()))


def wrapper():
    ws = torch.empty(256 * 1024 * 1024, dtype=torch.uint8, device=DEV)
    return flashinfer.BatchPrefillWithRaggedKVCacheWrapper(ws, kv_layout="NHD")


indptr = torch.tensor([0, SEQ], dtype=torch.int32, device=DEV)
capped, uncapped = reference(apply_cap=True), reference(apply_cap=False)

# 1) cap at plan() -- the documented way
w = wrapper()
w.plan(
    indptr,
    indptr,
    NH,
    NKV,
    HD,
    causal=True,
    sm_scale=SM_SCALE,
    logits_soft_cap=CAP,
    q_data_type=torch.bfloat16,
    kv_data_type=torch.bfloat16,
)
at_plan = w.run(q, k, v)

# 2) cap at run() -- rejected outright
w = wrapper()
w.plan(
    indptr, indptr, NH, NKV, HD, causal=True, sm_scale=SM_SCALE, q_data_type=torch.bfloat16, kv_data_type=torch.bfloat16
)
try:
    w.run(q, k, v, logits_soft_cap=CAP)
except TypeError as exc:
    print(f"run(): {exc}")

# 3) cap at the legacy forward() -- what the FlashInfer backend does
w = wrapper()
w.begin_forward(indptr, indptr, NH, NKV, HD, q_data_type=torch.bfloat16)
at_forward = w.forward(q, k, v, causal=True, sm_scale=SM_SCALE, logits_soft_cap=CAP)

for name, out in (("cap at plan()", at_plan), ("cap at legacy forward()", at_forward)):
    print(f"{name:<26} vs capped {cos(out, capped):.6f} | vs uncapped {cos(out, uncapped):.6f}")
```

```
cap at plan()              vs capped 0.999999 | vs uncapped 0.593867
run(): BatchPrefillWithRaggedKVCacheWrapper.run() got an unexpected keyword argument 'logits_soft_cap'
cap at legacy forward()    vs capped 0.593862 | vs uncapped 1.000000
```

The last line is the bug: cosine **1.000000** against the *uncapped* reference. The two references differ
at cos 0.5939, so this is not a tolerance question. `causal` and `sm_scale` passed to that same
`forward()` call *are* honored — only the cap is dropped, because it alone is compiled in at plan time.
Identical results on `flashinfer` 0.6.14 (torch 2.11) and 0.6.15.post1 (torch 2.13).

### 2. Same for `window_left` on the ragged path

Appending this to the script above (reduce the `* 12` gain to `* 2` so the cap is not what dominates):

```python
def windowed_reference(window_left):
    kf = k.float().repeat_interleave(NH // NKV, 1)
    vf = v.float().repeat_interleave(NH // NKV, 1)
    scores = torch.einsum("qhd,khd->hqk", q.float(), kf) * SM_SCALE
    idx = torch.arange(SEQ, device=DEV)
    keep = (idx[None, :] <= idx[:, None]) & (idx[None, :] >= idx[:, None] - window_left)
    return torch.einsum("hqk,khd->qhd", scores.masked_fill(~keep, float("-inf")).softmax(-1), vf)


w = wrapper()
w.begin_forward(indptr, indptr, NH, NKV, HD, q_data_type=torch.bfloat16)
out = w.forward(q, k, v, causal=True, sm_scale=SM_SCALE, window_left=8)
print(cos(out, windowed_reference(8)), cos(out, reference(apply_cap=False)))
```

```
window at plan()                 vs windowed 0.999999 | vs full 0.514832
window at legacy forward() only  vs windowed 0.514825 | vs full 0.999999
```

### 3. Model level: `gemma-2-27b`, `flashinfer` vs `triton`

```bash
python3 -m sglang.launch_server --model-path google/gemma-2-27b --dtype bfloat16 \
  --attention-backend flashinfer --disable-cuda-graph --disable-radix-cache
python3 -m sglang.launch_server --model-path google/gemma-2-27b --dtype bfloat16 \
  --attention-backend triton --disable-cuda-graph --disable-radix-cache
```

`google/gemma-2-27b` (`attn_logit_softcapping=50.0`, 46 layers, bf16), one 12-token prefill. Per-layer
module outputs captured with `nn.Module` forward hooks inside the scheduler process, compared against the
same forward in HF `transformers` (`attn_implementation="eager"`, same dtype). Cosine vs that reference:

| captured tensor | `flashinfer` | `triton` |
| --- | --- | --- |
| layer 0 output | 0.999999 | 0.999999 |
| layer 5 output | 0.999981 | 0.999999 |
| layer 12 output | 0.999513 | 0.999998 |
| layer 23 output | 0.999073 | 0.999990 |
| layer 45 output | **0.947087** | 0.999787 |
| layer 12 attention output | **0.921716** | 0.999723 |
| layer 45 attention output | **0.929931** | 0.999690 |
| layer 45 MLP output | **0.931770** | 0.999685 |

Re-running the HF reference with softcapping forced off moves it *onto* the FlashInfer output, which is
what identifies the cap rather than precision as the difference:

| captured tensor | capped HF vs FlashInfer | **uncapped** HF vs FlashInfer |
| --- | --- | --- |
| layer 5 attention output | 0.993836 | 0.999938 |
| layer 12 attention output | 0.921716 | 0.999238 |
| layer 45 attention output | 0.929931 | 0.995482 |
| layer 45 output | 0.947087 | 0.994924 |

### Why only some checkpoints notice

`tanh` capping is a no-op while `|logit| << cap`, so a model only diverges once its attention logits
approach the cap. Among gemma-2, `27b` is the one that does: its `query_pre_attn_scalar` (144) differs
from `head_dim` (128), so its pre-softmax logits are scaled up relative to `2b`/`9b`
(`query_pre_attn_scalar = 256 = head_dim`), which agree at cosine ≈ 1.0 on the same backend. `grok-1`
(cap 30) is in the same position, which is consistent with the GSM8K gap in flashinfer#1193.

### Suggested fix

Pass `logits_soft_cap` (and `window_left` on the ragged path) to `begin_forward()` / `plan()`, as the
paged wrapper already does for the window at :2185. Since the cap is per layer while planning is per
forward, that means planning per distinct cap value, or planning with the cap the model's layers request —
in practice one value per model for both affected models today.

Separately worth asking FlashInfer for: `forward()` should raise (or warn) when handed a parameter only
`plan()` can honor. `run()` already raises `TypeError`, which is how this surfaced.

Until this is fixed, `--attention-backend triton` produces correct output for capped models.

### Environment

`python3 -m sglang.check_env` on `main` (`18e6c61`, sglang `0.5.17.dev721`) with the torch, FlashInfer and
sgl-kernel versions it pins:

```
Python: 3.11.10 (main, Sep  7 2024, 18:35:41) [GCC 11.4.0]
CUDA available: True
GPU 0: NVIDIA A100 80GB PCIe
GPU 0 Compute Capability: 8.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.4, V12.4.131
CUDA Driver Version: 550.127.05
PyTorch: 2.13.0+cu130
sglang: 0.5.17.dev721+g18e6c61c2
sglang-kernel: 0.4.6.post1
flashinfer_python: 0.6.15.post1
flashinfer_cubin: Module Not Found
flashinfer_jit_cache: Module Not Found
triton: 3.7.1
transformers: 5.12.1
torchao: 0.17.0
numpy: 2.3.5
aiohttp: 3.14.3
fastapi: 0.141.1
huggingface_hub: 1.26.1
interegular: 0.3.3
modelscope: 1.39.1
orjson: 3.11.9
outlines: 0.1.11
packaging: 26.3
psutil: 7.2.2
pydantic: 2.13.4
python-multipart: 0.0.32
pyzmq: 27.1.0
uvicorn: 0.52.1
uvloop: 0.22.1
vllm: Module Not Found
xgrammar: 0.2.1
openai: 2.6.1
tiktoken: 0.13.0
anthropic: 0.120.2
litellm: Module Not Found
torchcodec: 0.15.0
NVIDIA Topology:
GPU0    NIC0    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      NODE    32-63,96-127    1               N/A
NIC0    NODE     X

ulimit soft: 1048576
```

The tables above were measured on the 0.5.16 release (`sglang-kernel` 0.4.5, `flashinfer_python` 0.6.14,
torch 2.11.0+cu130, same GPU and driver); the `main` environment above reproduces the same cosine values.
